File: PAT (Advanced Level) Practice/1178_File Path (25).cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>
#include <sstream>

using namespace std;

unordered_map<string, string> BuildChildParentMap(int n) {
    unordered_map<string, string> child_parent_map;
    stack<string> s;
    string line, id;

    for (int i = 0; i < n; ++i) {
        getline(cin, line);
        int depth = line.find_first_not_of(' ');
        id = line.substr(depth);
        while (s.size() > depth) s.pop();
        if (!s.empty()) child_parent_map[id] = s.top();
        s.push(id);
    }
    return child_parent_map;
}

// Function to trace back the path from a node to the root
string GetPath(const string& id, const unordered_map<string, string>& child_parent_map) {
    vector<string> path;
    string cur = id;

    while (child_parent_map.find(cur) != child_parent_map.end()) {
        path.push_back(cur);
        cur = child_parent_map.at(cur);
    }
    path.push_back("0000"); // Root ID

    string path_str;
    for (auto it = path.rbegin(); it != path.rend(); ++it) {
        if (it != path.rbegin()) path_str += "->";
        path_str += *it;
    }

    return path_str;
}

int main() {
    ios::sync_with_stdio(false);
    int n;
    cin >> n;
    cin.ignore(); // Ignore the newline after reading n

    auto child_parent_map = BuildChildParentMap(n);

    string queries_line;
    getline(cin, queries_line);
    istringstream iss(queries_line);
    string query;
    iss >> query; // Skip the first part, which indicates the number of queries

    while (iss >> query) {
        if (child_parent_map.find(query) != child_parent_map.end() || query == "0000") cout << GetPath(query, child_parent_map) << endl;
        else cout << "Error: " << query << " is not found." << endl;
    }

    return 0;
}

File: PAT (Advanced Level) Practice/1178_File Path (25).py
# Dictionary to store child -> parent relationships
parent_dict = {}


def build_tree(lines):
    # Stack to keep track of the current path
    stack = []
    # Parse the tree structure
    for line in lines:
        # Count the leading spaces to determine the depth
        depth = len(line) - len(line.lstrip(" "))
        # Get the ID without leading spaces
        id = line.strip()
        # If the stack is longer than the depth, pop elements
        while len(stack) > depth:
            stack.pop()
        # If the stack is not empty, the last element is the parent
        if stack:
            parent_dict[id] = stack[-1]
        # Push the current id onto the stack
        stack.append(id)


# Function to trace back the path from a node to the root
def get_path(id):
    path = []
    while id in parent_dict:
        path.append(id)
        id = parent_dict[id]
    path.append("0000")  # Root ID
    return "->".join(reversed(path))


n = int(input())
lines = []
for i in range(n):
    lines.append(input())
build_tree(lines)

queries = input().split()[1:]
# Process the queries
for query in queries:
    # Check if the ID exists in the tree
    if query in parent_dict or query == "0000":
        print(get_path(query))
    else:
        print(f"Error: {query} is not found.")
